core/src/main/scala/com/databricks/labs/remorph/preprocessors/Processor.scala

```diff
@@ -0,0 +1,10 @@
+package com.databricks.labs.remorph.preprocessors
+
+import com.databricks.labs.remorph._
+import org.antlr.v4.runtime.{CharStream, Lexer}
+
+trait Processor extends TransformationConstructors {
+  protected def createLexer(input: CharStream): Lexer
+  def pre(input: PreProcessing): Transformation[Parsing]
```

Comment:
It feels a bit odd for this method to return (a `Transformation` whose output is) a `Parsing` phase.

It is likely that what we want to express here is that `pre` it a `Transformation` that takes a `String` and returns a `String`, while also changing the `Phase` from `PreProcessing` to `Parsing`.

That would be better expressed as (inside `Processor` trait, which would then have a good reason for mixing in `TransformationConstructors`):

```scala
final def pre: Transformation[String] = {
  get
  .flatMap{
    case PerProcessing(input) => 
       val preprocessed  = preprocess(input)
      update{
         case _ => Parsing(preprocessed)
       }
    case _ => lift(KoResult(// pre shouldn't be called outside PerProcessing phase)
}

def preprocess(input: String): String
```

Reply:
I also wondered about this. But then I wondered if it was just that I got used to Parsing as being the start point. Let's have a chat aboiut this.

Reply:
Well, we should have changed `PlanParser#parse` the same way I suggested above for `Processor#pre` (getting rid of the `input` parameter and instead pull it from the state) long ago. It just hasn't happened yet, unfortunately (my bad, probably).